man/xtb.1.adoc
XTB(1)
======
:doctype: manpage

NAME
----

xtb - performs semiempirical quantummechanical calculations,
      for version 6.0 and newer

SYNOPSIS
--------

*xtb* ['OPTIONS'] 'FILE' ['OPTIONS']

DESCRIPTION
-----------

The `xtb(1)` program performs semiempirical quantummechanical calculations.
The underlying effective Hamiltonian is derived from density functional
tight binding (DFTB). This implementation of the xTB Hamiltonian is currently
compatible with the zeroth, first and second level parametrisation for
geometries, frequencies and non-covalent interactions (GFN)
as well as with the ionisation potential and
electron affinity (IPEA) parametrisation of the GFN1 Hamiltonian.
The generalized born (GB) model with solvent accessable surface area (SASA)
is also available in this version.
Ground state calculations for the simplified Tamm-Dancoff approximation (sTDA)
with the vTB model are currently not implemented.

GEOMETRY INPUT
~~~~~~~~~~~~~~

The wide variety of input formats for the geometry are supported by using
the mctc-lib. Supported formats are:

- Xmol/xyz files (xyz, log)
- Turbomole's coord, riper's periodic coord (tmol, coord)
- DFTB+ genFormat geometry inputs as cluster, supercell or fractional (gen)
- VASP's POSCAR/CONTCAR input files (vasp, poscar, contcar)
- Protein Database files, only single files (pdb)
- Connection table files, molfile (mol) and structure data format (sdf)
- Gaussian's external program input (ein)
- JSON input with `qcschema_molecule` or `qcschema_input` structure (json)
- FHI-AIMS' input files (geometry.in)
- Q-Chem molecule block inputs (qchem)

For a full list visit: https://grimme-lab.github.io/mctc-lib/page/index.html

`xtb(1)` reads additionally `.CHRG` and `.UHF` files if present.

INPUT SOURCES
-------------
`xtb(1)` gets its information from different sources. The one with highest
priority is the commandline with all allowed flags and arguments described below.
The secondary source is the `xcontrol(7)` system, which can in principle
use as many input files as wished. The `xcontrol(7)` system is the successor
of the set-block as present in version 5.8.2 and earlier. This implementation
of `xtb(1)` reads the `xcontrol(7)` from two of three possible sources,
the local xcontrol file or the 'FILE' used to specify the geometry
and the global configuration file found in the `XTBPATH`.

OPTIONS
-------

*-c, --chrg* 'INT':: 
    specify molecular charge as 'INT', overrides `.CHRG` file and `xcontrol` option

*-c, --chrg* 'INT:INT':: 
    specify charges for 'inner region:outer region' for `oniom` calculation, overrides `.CHRG` file and `xcontrol` option

*-u, --uhf* 'INT'::
    specify number of unpaired electrons as 'INT', overrides `.UHF` file and `xcontrol` option
    (default: 0 (singlet) for even electron number, 1 (doublet) for odd electron number)

*--gfn* 'INT'::
    specify parametrisation of GFN-xTB (default = 2)

*--gfnff, --gff* ::
    specify parametrisation of GFN-FF

*--tblite* ::
    use tblite library as implementation for xTB
    
*--ceh <grad> [REAL]* ::
      calculate CEH (Charge-Extended Hückel model) charges and write them to the ceh.charges file,
      optionally, calculate numerical gradients and write them to the ceh.charges.numgrad file
      with an adjustable step size for the numerical gradients.

*--ptb* ::
    performs single-point calculation with the density tight-binding method PTB.
    Provides electronic structure and properties, such as, e.g., atomic charges, bond orders, and dipole moments,
    but does not provide any energy-related properties, such as, e.g., total energy, nuclear gradients, or vibrational frequencies.

*--spinpol*::
    enables spin-polarization for xTB methods (tblite required)

*--oniom* 'METHOD' 'LIST'::
    use subtractive embedding via ONIOM method. 'METHOD' is given as `high:low`
    where `high` can be 'orca', 'turbomole', 'gfn2', 'gfn1', or 'gfnff' and
    `low` can be 'gfn2', 'gfn1', or 'gfnff'.
    The inner region is given as comma-separated indices directly in the commandline
    or in a file with each index on a separate line.
    Combine with --tblite to evaluate 'gfn2' and 'gfn1' with the tblite library.

*--efield* 'REAL,REAL,REAL'::
    static electric field in Cartesian coordinates, overrides `.EFIELD` file,
    works only via tblite for the xTB Hamiltonians, or with GFN-FF and PTB.

*--etemp, --temp* 'REAL'::
    electronic temperature for SCC (default = 300K)

*--esp* ::
    calculate electrostatic potential on VdW-grid

*--stm* ::
    calculate STM image

*-a, --acc* 'REAL'::
    accuracy for SCC calculation, lower is better (default = 1.0),
    overwrites also tighter defaults for optimization and hessians

*--iterations, --maxiterations* 'INT'::
    maximum number of SCC iterations per single point calculation (default = 250)

*--vparam* 'FILE'::
    Parameter file for xTB calculation

*--alpb* 'SOLVENT' ['STATE']::
    analytical linearized Poisson-Boltzmann (ALPB) model,
    available solvents are 'acetone', 'acetonitrile', 'aniline', 'benzaldehyde',
    'benzene', 'ch2cl2', 'chcl3', 'cs2', 'dioxane', 'dmf', 'dmso', 'ether',
    'ethylacetate', 'furane', 'hexandecane', 'hexane', 'methanol', 'nitromethane',
    'octanol', 'woctanol', 'phenol', 'toluene', 'thf', 'water'.
    The solvent input is not case-sensitive.
    The Gsolv reference state can be chosen as 'reference', 'bar1M', or 'gsolv' (default).

*-g, --gbsa* 'SOLVENT' ['STATE']::
    generalized born (GB) model with solvent accessable surface (SASA) model,
    available solvents are 'acetone', 'acetonitrile', 'benzene' (only GFN1-xTB),
    'CH2Cl2', 'CHCl3', 'CS2', 'DMF' (only GFN2-xTB), 'DMSO', 'ether', 'H2O',
    'methanol', 'n-hexane' (only GFN2-xTB), 'THF' and 'toluene'.
    The solvent input is not case-sensitive.
    The Gsolv reference state can be chosen as 'reference', 'bar1M', or 'gsolv' (default).

*--gbe* 'SOLVENT/EPSILON'::
    generalized Born for finite epsilon (GBe) solvation model (tblite required),
    includes only the electrostatic solvation contribution (no SASA model),
    available solvents are all solvents that are available for alpb.
    Additionally, the dielectric constant can be set manually.

*--gb* 'SOLVENT/EPSILON'::
    generalized Born (GB) solvation model (tblite required),
    includes only the electrostatic solvation contribution (no SASA model),
    available solvents are all solvents that are available for alpb.
    Additionally, the dielectric constant can be set manually.

*--cosmo* 'SOLVENT/EPSILON'::
    domain decomposition conductor-like screening model (ddCOSMO)
    available solvents are all solvents that are available for alpb.
    Additionally, the dielectric constant can be set manually or an ideal conductor
    can be chosen by setting epsilon to infinity.
    Available via tblite or in the legacy version in xtb lacking density response
    to the continuum polarization (non-variational).

*--cpcm* 'SOLVENT/EPSILON'::
    domain decomposition conductor-like polarizable continuum solvation model (ddCPCM),
    available solvents are all solvents that are available for alpb.
    Additionally, the dielectric constant can be set manually or an ideal conductor
    can be chosen by setting epsilon to infinity.
    Only available via tblite.

*--pcm* 'SOLVENT/EPSILON'::
    domain decomposition polarizable continuum model (ddPCM),
    available solvents are all solvents that are available for alpb.
    Additionally, the dielectric constant can be set manually or an ideal conductor
    can be chosen by setting epsilon to infinity.
    Only available via tblite.

*--tmcosmo* 'SOLVENT/EPSILON'::
    same as --cosmo, but uses TM convention for writing the .cosmo files.

*--cpcmx* 'SOLVENT'::
    extended conduction-like polarizable continuum solvation model (CPCM-X),
    available solvents are all solvents included in the Minnesota Solvation Database.

*--cma* ::
    shifts molecule to center of mass and transforms cartesian coordinates
    into the coordinate system of the principle axis (not affected by
    `isotopes'-file).

*--pop*::
    requests printout of Mulliken population analysis

*--molden*::
    requests printout of molden file (name: molden.input), format differs with --tblite

*--alpha* ::
    requests the extension of electrical properties to static molecular dipole polarizabilities

*--raman* ::
    requests Raman spectrum calculation via combination of GFN2-xTB and PTB
    using the temperature 'REAL' (default 298.15 K) and the wavelength of 
    the incident laser which must be given in nm 'REAL' (default 514 nm)
    
*--dipole*::
    requests dipole printout

*--quadrupole*::
    requests quadrupole printout

*--wbo*::
    requests Wiberg bond order printout

*--lmo*::
    requests localization of orbitals

*--fod*::
    requests FOD calculation

*--cycles* 'INT'::
    maximum number of optimization cycles

RUNTYPS
~~~~~~~
NOTE: You can only select *one* runtyp, only the first runtyp will be used
      from the program, use implemented composite runtyps to perform several
      operations at once.

*--scc, --sp*::
    performs a single point calculation

*--vip*::
    performs calculation of ionisation potential.
    This needs the .param_ipea.xtb parameters
    and a GFN1 Hamiltonian.

*--vea*::
    performs calculation of electron affinity.
    This needs the .param_ipea.xtb parameters
    and a GFN1 Hamiltonian.

*--vipea*::
    performs calculation of electron affinity and ionisation potential.
    This needs the .param_ipea.xtb parameters
    and a GFN1 Hamiltonian.

*--vfukui*::
    performs calculation of Fukui indices.

*--vomega*::
    performs calculation of electrophilicity index.
    This needs the .param_ipea.xtb parameters
    and a GFN1 Hamiltonian.

*--grad*::
    performs a gradient calculation

*-o, --opt* ['LEVEL']::
    call `ancopt(3)` to perform a geometry optimization,
    levels from crude, sloppy, loose, normal (default), tight, verytight
    to extreme can be chosen, level tightens also the default accuracy setting 
    (0.5 tight, and 0.2 verytight/extreme, overwritten by --acc flag)

*--hess*::
    perform a numerical hessian calculation on input geometry,
    tightens default accuracy setting to 0.2 (overwritten by --acc flag)

*--ohess* ['LEVEL']::
    perform a numerical hessian calculation on an `ancopt(3)` optimized
    geometry, tightens default accuracy setting to 0.2 (overwritten by --acc flag)

*--bhess* ['LEVEL']::
    perform a biased numerical hessian calculation on an `ancopt(3)` optimized
    geometry, tightens default accuracy setting to 0.2 (overwritten by --acc flag)

    Hessian runs automatically perform a compliance analysis for nonperiodic
    systems with more than one atom and no frozen atoms.

*--o1nh*::
    perform the numerical hessian calculation using the ODLR approximation (O1NumHess)

*--imagmin* 'REAL'::
    magnitudes below this value (cm⁻¹) are ignored in the O1NumHess
    imaginary frequency repair (default 5).

*--imagmax* 'REAL'::
    deepest imaginary frequency (cm⁻¹) considered in the O1NumHess
    imaginary frequency repair (default 200).

*--ithr* 'REAL'::
    imaginary frequencies between this cutoff (cm⁻¹) and zero are treated
    as vibrations in the RRHO calculation (default 20).

    Hessian runs automatically perform a compliance analysis for nonperiodic
    systems with more than one atom and no frozen atoms.

*--md*::
    molecular dynamics simulation on start geometry

*--metadyn* ['int']::
    meta dynamics simulation on start geometry, saving 'int' snapshots
    of the trajectory to bias the simulation

*--omd*::
    molecular dynamics simulation on `ancopt(3)` optimized geometry,
    a loose optimization level will be chosen

*--metaopt* ['LEVEL']::
    call `ancopt(3)` to perform a geometry optimization,
    then try to find other minimas by meta dynamics, 
    level tightens also the default accuracy setting
    (0.5 tight, and 0.2 verytight/extreme, overwritten by --acc flag)

*--path* ['FILE']::
    use meta dynamics to calculate a path from the input geometry
    to the given product structure

*--reactor*::
    experimental

*--modef* 'INT'::
    modefollowing algorithm. 'INT' specifies the mode that should be
    used for the modefollowing.

*--dipro* ['REAL']:: 
    the dimer projection method for the calculation of electronic coupling integrals between two fragments.
    'REAL' sets the threshold for nearly degenerate orbitals to still be considered (default = 0.1 eV).
    

GENERAL
~~~~~~~
*-I, --input* 'FILE'::
     use 'FILE' as input source for `xcontrol(7)` instructions

*--namespace* 'STRING'::
     give this `xtb(1)` run a namespace. All files, even temporary
     ones, will be named according to 'STRING' (might not work everywhere).

*--[no]copy*::
     copies the `xcontrol` file at startup (default = true)

*--[no]restart*::
     restarts calculation from `xtbrestart` (default = true)

*-P, --parallel* 'INT'::
     number of parallel processes

*--define*::
     performs automatic check of input and terminate

*--json*::
     write xtbout.json file

*--citation*::
     print citation and terminate

*--license*::
     print license and terminate

*-v, --verbose*::

python/xtb.py
#!/usr/bin/env python

from __future__ import print_function

import numpy as np
import ase
import ctypes

from ctypes import cdll, Structure, c_double, c_int, c_bool, POINTER, c_char_p

from ase import Atoms
from ase.calculators.calculator import Calculator, all_changes
from ase.units import Hartree, Bohr

class SCC_options(Structure):
    _fields_ = [('prlevel', c_int),
                ('parallel',c_int),
                ('acc',     c_double),
                ('etemp',   c_double),
                ('grad',    c_bool),
                ('restart', c_bool),
                ('maxiter', c_int),
                ('solvent', ctypes.c_char*20)]

class PEEQ_options(Structure):
    _fields_ = [('prlevel', c_int),
                ('parallel',c_int),
                ('acc',     c_double),
                ('etemp',   c_double),
                ('grad',    c_bool),
                ('ccm',     c_bool)]

class XTB(Calculator):
    implemented_properties = ['energy', 'free_energy', 'forces']
    command = None

    def __init__(self, charge = None, accuracy = None, temperature = None,
            print_level = None,
            restart=None, ignore_bad_restart_file=False,
            label="saw", atoms=None, command=None, debug=False, **kwargs):
        """Construct the xtb-calculator object."""

        self._debug = debug
        self._lib = None
        self.label = None
        self.parameters = None
        self.results = None
        self.atoms = None

        if charge is not None:
            self.charge = charge
        else:
            self.charge = 0.0

        if accuracy is not None:
            self.accuracy = accuracy
        else:
            self.accuracy = 1.0

        if temperature is not None:
            self.temperature = temperature
        else:
            self.temperature = 300.0

        if print_level is not None:
            self.print_level = print_level
        else:
            self.print_level = 2

        if command is not None:
            self.command = command
        else:
            self.command = 'xtb' # default

        Calculator.__init__(self, restart, ignore_bad_restart_file,
                label, atoms, **kwargs)

        if restart is not None:
            try:
                self.read(restart)
            except:
                if ignore_bad_restart_file:
                    self.reset()
                else:
                    raise

    def set(self, **kwargs):
        """Set parameters like set(key1=value1, key2=value2, ...)."""
        changed_parameters = Calculator.set(self, **kwargs)
        if changed_parameters:
            self.reset()

    def write(self, label):
        """Write atoms, parameters and calculated results into restart files."""
        if self._debug:
            print("Writting restart to: ", label)
        self.atoms.write(label + '_restart.traj')
        self.parameters.write(label + '_params.ase')
        open(label + '_results.ase', 'w').write(repr(self.results))

    def read(self, label):
        """Read atoms, parameters and calculated results from restart files."""
        self.atoms = ase.io.read(label + '_restart.traj')
        self.parameters = Parameters.read(label + '_params.ase')
        results_txt = open(label + '_results.ase').read()
        self.results = eval(results_txt, {'array': np.array})

    def _open_lib(self):
        # load xtb-library
        self._lib = cdll.LoadLibrary('libxtb.so')

        c_int_p = POINTER(c_int)
        c_bool_p = POINTER(c_bool)
        c_double_p = POINTER(c_double)

        # define periodic GFN0-xTB interface
        self._lib.GFN0_PBC_calculation.argtypes = [c_int_p, c_int_p,
            c_double_p, c_double_p, c_double_p,
            c_bool_p, POINTER(PEEQ_options), c_char_p,
            c_double_p, c_double_p, c_double_p]

        # define GFN0-xTB interface
        self._lib.GFN0_calculation.argtypes = [c_int_p, c_int_p,
            c_double_p, c_double_p, POINTER(PEEQ_options), c_char_p,
            c_double_p, c_double_p]

        # define GFN1-xTB interface
        self._lib.GFN1_calculation.argtypes = [c_int_p, c_int_p,
            c_double_p, c_double_p, POINTER(SCC_options), c_char_p,
            c_double_p, c_double_p]

        # define GFN2-xTB interface
        self._lib.GFN2_calculation.argtypes = [c_int_p, c_int_p,
            c_double_p, c_double_p, POINTER(SCC_options), c_char_p,
            c_double_p, c_double_p, c_double_p, c_double_p, c_double_p,
            c_double_p, c_double_p]

    def calculate(self, atoms=None, properties=None, system_changes=all_changes):
        """Do the calculation."""

        if not properties:
            properties =  ['energy']
        Calculator.calculate(self, atoms, properties, system_changes)

class GFN0(XTB):
    implemented_properties = ['energy', 'free_energy', 'forces', 'stress']
    command = None

    def __init__(self, charge = None, accuracy = None, temperature = None,
            print_level = None,
            restart=None, ignore_bad_restart_file=False,
            label="saw", atoms=None, command=None, debug=False, **kwargs):
        """Construct the xtb-calculator object."""

        XTB.__init__(self, charge, accuracy, temperature, print_level,
                restart, ignore_bad_restart_file, label, atoms, **kwargs)


    def calculate(self, atoms=None, properties=None, system_changes=all_changes):
        """Do the calculation."""

        if not properties:
            properties =  ['energy']
        XTB.calculate(self, atoms, properties, system_changes)

        if self._debug:
            print("system_changes:", system_changes)

        if self._lib is None:
            self._open_lib()

        natoms  = c_int(atoms.get_number_of_atoms())
        attyp   = np.array(atoms.get_atomic_numbers(), dtype=np.int32)
        attyp_p = attyp.ctypes.data_as(POINTER(c_int))
        coord   = atoms.get_positions()/Bohr # from Angstrom to Bohr
        coord_p = coord.ctypes.data_as(POINTER(c_double))
        charge  = c_double(self.charge)

        energy   = c_double(0.0)
        gradient = np.zeros((3,natoms.value),order='F')
        grad_p   = gradient.ctypes.data_as(POINTER(c_double))

        opt = PEEQ_options(c_int(self.print_level),
                           c_int(0), # parallel -> automatic
                           c_double(self.accuracy),
                           c_double(self.temperature),
                           c_bool(True), # gradient -> always
                           c_bool(True)) # CCM -> yes

        outfile = "gfn0.out".encode('utf-8')

        stat = self._lib.GFN0_calculation(natoms,attyp_p,charge,
                coord_p,opt,outfile,energy,grad_p)
        # in case Fortran behaves we find a useful return value
        if (stat != 0):
            raise Exception("xtb terminated in error.")

        self.results['energy'] = energy.value * Hartree
        self.results['free_energy'] = energy.value * Hartree
        self.results['forces'] = - gradient.T * Hartree / Bohr

class GFN1(XTB):
    implemented_properties = ['energy', 'free_energy', 'forces']
    command = None

    def __init__(self, charge = None, accuracy = None, temperature = None,
            max_iterations = None, print_level = None, solvent = None,
            restart=None, ignore_bad_restart_file=False,
            label="saw", atoms=None, command=None, debug=False, **kwargs):
        """Construct the GFN2-calculator object."""

        if max_iterations is not None:
            self.max_iterations = max_iterations
        else:
            self.max_iterations = 250

        if solvent is not None:
            self.solvent = solvent
        else:
            self.solvent = "none"

        XTB.__init__(self, charge, accuracy, temperature, print_level,
                restart, ignore_bad_restart_file, label, atoms, **kwargs)

    def calculate(self, atoms=None, properties=None, system_changes=all_changes):
        """Do the calculation."""

        if not properties:
            properties =  ['energy']
        XTB.calculate(self, atoms, properties, system_changes)

        if self._debug:
            print("system_changes:", system_changes)

        if self._lib is None:
            self._open_lib()

        natoms  = c_int(atoms.get_number_of_atoms())
        attyp   = np.array(atoms.get_atomic_numbers(), dtype=np.int32)
        attyp_p = attyp.ctypes.data_as(POINTER(c_int))
        coord   = atoms.get_positions()/Bohr # from Angstrom to Bohr
        coord_p = coord.ctypes.data_as(POINTER(c_double))
        charge  = c_double(self.charge)

        energy   = c_double(0.0)
        gradient = np.zeros((3,natoms.value),order='F')
        grad_p   = gradient.ctypes.data_as(POINTER(c_double))

        opt = SCC_options(c_int(self.print_level),
                          c_int(0), # parallel -> automatic
                          c_double(self.accuracy),
                          c_double(self.temperature),
                          c_bool(True),  # gradient -> always
                          c_bool(False), # restart -> no
                          c_int(self.max_iterations),
                          self.solvent.encode('utf-8'))

        outfile = "gfn1.out".encode('utf-8')

        stat = self._lib.GFN1_calculation(natoms,attyp_p,charge,
                coord_p,opt,outfile,energy,grad_p)
        # in case Fortran behaves we find a useful return value
        if (stat != 0):
            raise Exception("xtb terminated in error.")


        self.results['energy'] = energy.value * Hartree
        self.results['free_energy'] = energy.value * Hartree
        self.results['forces'] = - gradient.T * Hartree / Bohr

class GFN2(XTB):
    implemented_properties = ['energy', 'free_energy', 'forces', 'dipole',
            'charges', 'dipoles', 'quadrupoles', 'wiberg']
    command = None

    def __init__(self, charge = None, accuracy = None, temperature = None,
            max_iterations = None, print_level = None, solvent = None,
            restart=None, ignore_bad_restart_file=False,
            label="saw", atoms=None, command=None, debug=False, **kwargs):
        """Construct the GFN2-calculator object."""

        if max_iterations is not None:
            self.max_iterations = max_iterations
        else:
            self.max_iterations = 250

        if solvent is not None:
            self.solvent = solvent
        else:
            self.solvent = "none"

        XTB.__init__(self, charge, accuracy, temperature, print_level,
                restart, ignore_bad_restart_file, label, atoms, **kwargs)

    def calculate(self, atoms=None, properties=None, system_changes=all_changes):
        """Do the calculation."""

        if not properties:
            properties =  ['energy']
        XTB.calculate(self, atoms, properties, system_changes)

        if self._debug:
            print("system_changes:", system_changes)

        if self._lib is None:
            self._open_lib()

        natoms  = c_int(atoms.get_number_of_atoms())
        attyp   = np.array(atoms.get_atomic_numbers(), dtype=np.int32)
        attyp_p = attyp.ctypes.data_as(POINTER(c_int))
        coord   = atoms.get_positions()/Bohr # from Angstrom to Bohr
        coord_p = coord.ctypes.data_as(POINTER(c_double))
        charge  = c_double(self.charge)
        self.results['charges']     = np.zeros(natoms.value)
        self.results['dipoles']     = np.zeros((3,natoms.value),order='F')
        self.results['quadrupoles'] = np.zeros((6,natoms.value),order='F')
        self.results['dipole'] = np.zeros(3)
        self.results['wiberg'] = np.zeros((natoms.value,natoms.value),order='F')
        q_p      = self.results['charges'].ctypes.data_as(POINTER(c_double))
        dipm_p   = self.results['dipoles'].ctypes.data_as(POINTER(c_double))
        qp_p     = self.results['quadrupoles'].ctypes.data_as(POINTER(c_double))
        dipole_p = self.results['dipole'].ctypes.data_as(POINTER(c_double))
        wbo_p    = self.results['wiberg'].ctypes.data_as(POINTER(c_double))

        energy   = c_double(0.0)
        gradient = np.zeros((3,natoms.value),order='F')
        grad_p   = gradient.ctypes.data_as(POINTER(c_double))

        opt = SCC_options(c_int(self.print_level),
                          c_int(0), # parallel -> automatic
                          c_double(self.accuracy),
                          c_double(self.temperature),
                          c_bool(True), # gradient -> always
                          c_bool(True), # restart  -> if possible
                          c_int(self.max_iterations),
                          self.solvent.encode('utf-8'))

        outfile = "gfn2.out".encode('utf-8')

        stat = self._lib.GFN2_calculation(natoms,attyp_p,charge,
            coord_p,opt,outfile,energy,grad_p,dipole_p,q_p,dipm_p,qp_p,wbo_p)
        # in case Fortran behaves we find a useful return value
        if (stat != 0):
            raise Exception("xtb terminated in error.")

        self.results['energy'] = energy.value * Hartree
        #print("-> GFN2:", energy.value * Hartree)
        self.results['free_energy'] = energy.value * Hartree
        self.results['forces'] = - gradient.T * Hartree / Bohr

        # some unit conversion for the sake of consistency
        self.results['dipole']  = self.results['dipole']  * Bohr
        self.results['dipoles'] = self.results['dipoles'] * Bohr
        self.results['quadrupoles'] = self.results['quadrupoles'] * Bohr*Bohr

class GFN0_PBC(XTB):
    implemented_properties = ['energy', 'free_energy', 'forces', 'stress']
    command = None

    def __init__(self, charge = None, accuracy = None, temperature = None,
            print_level = None,
            restart=None, ignore_bad_restart_file=False,
            label="saw", atoms=None, command=None, debug=False, **kwargs):
        """Construct the xtb-calculator object."""

        XTB.__init__(self, charge, accuracy, temperature, print_level,
                restart, ignore_bad_restart_file, label, atoms, **kwargs)


    def calculate(self, atoms=None, properties=None, system_changes=all_changes):
        """Do the calculation."""

        if not properties:
            properties =  ['energy']
        XTB.calculate(self, atoms, properties, system_changes)

        if self._debug:
            print("system_changes:", system_changes)

        if self._lib is None:
            self._open_lib()

        natoms  = c_int(atoms.get_number_of_atoms())
        attyp   = np.array(atoms.get_atomic_numbers(), dtype=np.int32)
        attyp_p = attyp.ctypes.data_as(POINTER(c_int))
        coord   = atoms.get_positions()/Bohr # from Angstrom to Bohr
        coord_p = coord.ctypes.data_as(POINTER(c_double))
        lattice = atoms.get_cell()/Bohr # from Angstrom to Bohr
        dlat_p  = lattice.ctypes.data_as(POINTER(c_double))
        charge  = c_double(self.charge)

        energy   = c_double(0.0)
        gradient = np.zeros((3,natoms.value),order='F')
        grad_p   = gradient.ctypes.data_as(POINTER(c_double))
        gradlatt = np.zeros((3,3),order='F')
        glat_p   = gradlatt.ctypes.data_as(POINTER(c_double))

        opt = PEEQ_options(c_int(self.print_level),
                           c_int(0), # parallel -> automatic
                           c_double(self.accuracy),
                           c_double(self.temperature),
                           c_bool(True), # gradient -> always
                           c_bool(True)) # CCM -> yes

        #pbc = np.array([True,True,True], dtype=np.bool)
        pbc = atoms.get_pbc()
        pbc_p = pbc.ctypes.data_as(POINTER(c_bool))

        outfile = "gfn0.out".encode('utf-8')

        # try/except is not working on Fortran
        stat = self._lib.GFN0_PBC_calculation(natoms,attyp_p,charge,
                coord_p,dlat_p,pbc_p,opt,outfile,energy,grad_p,glat_p)
        # in case Fortran behaves we find a useful return value
        if (stat != 0):
            raise Exception("xtb terminated in error.")

        self.results['energy'] = energy.value * Hartree
        self.results['free_energy'] = energy.value * Hartree
        self.results['forces'] = - gradient.T * Hartree / Bohr
        # xtb returns the cell gradient, so we have to backtransform
        # taking into account that the stress tensor should be symmetric,
        # we interface F and C ordered arrays and have to transpose we
        # can simply drop all reshape and transpose steps by writing
        stress = np.dot(self.atoms.cell,
                gradlatt * Hartree / Bohr / self.atoms.get_volume())
        self.results['stress'] = stress.flat[[0, 4, 8, 5, 2, 1]]

if __name__ == "__main__":

    import argparse

    def get_cmd_args():
        parser = argparse.ArgumentParser(prog='xtb',
                description='Wrapper for xTB calculation.')

        parser.add_argument("-f","--format",dest="filetype",action="store",
                default=None,help="Format of input geometry (automatic)")

        parser.add_argument(dest="filename",action="store",
                help="Input geometry")

        parser.add_argument("-x","--method",dest="method",action="store",
                default='gfn2',help="SQM method for calculation (gfn2)")

        parser.add_argument("-c","--charge",dest="charge",action="store",
                type=int,default=0,help="Total charge (0)")

        parser.add_argument("--etemp",dest="etemp",action="store",
                type=float,default=300.0,help="Electronic temperature (300K)")

        parser.add_argument("--accuracy",dest="acc",action="store",
                type=float,default=1.0,help="Calculation accuracy (1.0)")

        parser.add_argument("--maxiter",dest="maxiter",action="store",
                type=int,default=250,help="Maximum number of SCC iterations (250)")

        parser.add_argument("--lcovb",dest="rnn",action="store",
                type=float,default=5.0,help="Longest covalent bond for preconditioning (5.0)")

        parser.add_argument("--econv",dest="econv",action="store",
                type=float,default=5.4423e-4,help="Convergence threshold for energy")

        parser.add_argument("--gconv",dest="fconv",action="store",
                type=float,default=3.88938e-05,help="Convergence threshold for forces")

        parser.add_argument("--linesearch",dest="armijo",action="store_true",
                help="Use line search")

        parser.add_argument("--precon",dest="precon",action="store_true",
                help="Use preconditioning")

        parser.add_argument("--optcell",dest="optcell",action="store_true",
                help="Relax cellparameters")

        parser.add_argument("--logfile",dest="logfile",action="store",
                default='-',help="File for logging information (STDOUT)")

        parser.add_argument("--trajectory",dest="trajectory",action="store",
                default='xtbopt.traj',help="File for trajectory output (xtbopt.traj)")

        parser.add_argument("--solvent",dest="solvent",action="store",
                default="none",help="Solvent for GBSA (none)")

        return parser.parse_args()

    import numpy
    import ase
    from ase.io import read, write
    from ase.units import *
    from ase.optimize.precon import Exp, PreconFIRE
    from ase.constraints import ExpCellFilter

    # overwrite convergence thresholds of PreconFIRE optimizer
    class PatchedOptimizer(PreconFIRE):

        def initialize(self):
            PreconFIRE.initialize(self)
            self.elast = None

        def run(self, steps=100000000, econv = 0.00054423,fconv = 3.88938e-05):
            self.econv = econv
            self.fconv = fconv
            PreconFIRE.run(self, 0.05, steps)

        def converged(self, forces=None):

            # get current energy and check for convergence in energy change
            ecurr = self.atoms.get_potential_energy()
            if self.elast is not None:
                ediff = ecurr - self.elast
                econverged = ediff < 0.0 and abs(ediff) < self.econv
            else:
                econverged = False
            # save potential energy
            self.elast = ecurr

            # check for convergence in forces
            if forces is None:
                forces = self.atoms.get_forces()
            fnorm = sqrt((forces**2).sum())
            fconverged = fnorm < self.fconv

            #print("--> norm(F):", fnorm, "converged?", fconverged)
            #print("--> energy :", ecurr, "converged?", econverged)

            return econverged and fconverged

    args = get_cmd_args()
    
    if args.filetype is not None:
        mol = ase.io.read(args.filename, format = str(args.filetype))
    else:
        mol = ase.io.read(args.filename)

    if args.method == 'gfn0':
        if mol.pbc.any():
            calc = GFN0_PBC(charge=args.charge,accuracy=args.acc,
                    temperature=args.etemp,print_level=2)
        else:
            calc = GFN0(charge=args.charge,accuracy=args.acc,temperature=args.etemp,
                    print_level=2)
    elif args.method == 'gfn1':
        if mol.pbc.any():
            raise Exception("GFN1-xTB is not available with PBC")
        else:
            calc = GFN1(charge=args.charge,accuracy=args.acc,temperature=args.etemp,
                    max_iterations=args.maxiter,print_level=2,solvent=args.solvent)
    elif args.method == 'gfn2':
        if mol.pbc.any():
            raise Exception("GFN2-xTB is not available with PBC")
        else:
            calc = GFN2(charge=args.charge,accuracy=args.acc,temperature=args.etemp,
                    max_iterations=args.maxiter,print_level=2,solvent=args.solvent)
    else:
        raise Exception("Method not implemented.")
    mol.set_calculator(calc)

    e = mol.get_potential_energy()

    print('Initial energy: eV, Eh',e, e/Hartree)

    if args.precon:
        precon = Exp(A=3, r_NN=args.rnn + 0.1, r_cut=args.rnn + 0.6)
    else:
        precon = None

    if args.optcell:
        sf = ExpCellFilter(mol)
        relax = PatchedOptimizer(sf, precon=precon,
                trajectory=args.trajectory, logfile=args.logfile, use_armijo=args.armijo)
    else:
        relax = PatchedOptimizer(mol, precon=precon,
                trajectory=args.trajectory, logfile=args.logfile, use_armijo=args.armijo)

    try:
        relax.run(econv=args.econv,fconv=args.fconv)
    except KeyboardInterrupt:
        print('User got impatient')
    except RuntimeError:
        print('Optimization terminated due to internal error')

    e = mol.get_potential_energy()
    print('Final energy: eV, Eh',e, e/Hartree)

    if (mol.pbc.any()):
        ase.io.write("xtbopt.POSCAR",mol,format='vasp')
    else:
        ase.io.write("xtbopt.xyz",mol,format='xyz')
